packages/dds/tree/src/util/utils.ts

```diff
@@ -79,6 +79,44 @@ export function makeArray<T>(size: number, filler: (index: number) => T): T[] {
 	return array;
 }
 
+/**
+ * Returns the last element of an array, or `undefined` if the array has no elements.
+ * @param array - The array to get the last element from.
+ * @remarks
+ * If the type of the array has been narrowed by e.g. {@link hasSome | hasSome(array)} or {@link hasOne | hasOne(array)} then the return type will be `T` rather than `T | undefined`.
+ */
+export function getLast<T>(array: readonly [T, ...T[]]): T;
+export function getLast<T>(array: { [index: number]: T; length: number }): T | undefined;
+export function getLast<T>(array: { [index: number]: T; length: number }): T | undefined {
+	return array[array.length - 1];
+}
+
+/**
+ * Returns true if and only if the given array has at least one element.
+ * @param array - The array to check.
+ * @remarks
+ * If `array` contains at least one element, its type will be narrowed and can benefit from improved typing from e.g. `array[0]` and {@link getLast | getLast(array)}.
+ * This is especially useful when "noUncheckedIndexedAccess" is enabled in the TypeScript compiler options, since the return type of `array[0]` will be `T` rather than `T | undefined`.
+ */
+export function hasSome<T>(array: T[]): array is [T, ...T[]];
+export function hasSome<T>(array: readonly T[]): array is readonly [T, ...T[]];
+export function hasSome<T>(array: readonly T[]): array is [T, ...T[]] {
+	return array.length > 0;
+}
+
+/**
+ * Returns true if and only if the given array has exactly one element.
```

Comment:
could wait, but i think a lot of these would be useful in client-utils so they can be used generally

Reply:
I think that's a great idea. [I've filed a work item to track this](https://dev.azure.com/fluidframework/internal/_workitems/edit/25251). We can use it as an opportunity to get more opinions on what the functions should be named too, since there are a lot of options (hasSome, hasElements, hasAtLeastOne, isNonEmpty, etc...) and I'm not sure they have the best name yet. I'll check in this as-is first since the improved typing it allows for SharedTreeBranchEvents is going to be useful for some imminent work in the tree package.

Reply:
for sure. this will all be internal, so we have more flexibility as well